eclair-core/src/main/scala/fr/acinq/eclair/payment/MultiPartPaymentHandler.scala

```diff
@@ -0,0 +1,143 @@
+    // If we receive htlcs after the multi-part payment has expired, we must fail them.
+    // The LocalPaymentHandler will create a new instance of MultiPartPaymentHandler to handle a new attempt.
+    case Event(MultiPartHtlc(_, htlc), PaymentFailed(_, failure, _)) =>
+      require(htlc.paymentHash == paymentHash, s"invalid payment hash (expected $paymentHash, received ${htlc.paymentHash}")
+      parent ! ExtraHtlcReceived(paymentHash, PendingPayment(htlc.id, PartialPayment(htlc.amountMsat, htlc.channelId), sender), Some(failure))
+      stay
+
+    case Event("ok", _) => stay
+  }
+
+  onTransition {
+    case _ -> PAYMENT_SUCCEEDED =>
+      cancelTimer(PaymentTimeout.toString)
+      nextStateData match {
+        case PaymentSucceeded(_, parts) =>
+          // We expect the parent actor to send us a PoisonPill after receiving this message.
+          parent ! MultiPartHtlcSucceeded(paymentHash, parts.reverse)
+        case d =>
+          log.error(s"unexpected payment success data ${d.getClass.getSimpleName}")
+      }
+    case _ -> PAYMENT_FAILED =>
+      cancelTimer(PaymentTimeout.toString)
+      nextStateData match {
+        case PaymentFailed(_, failure, parts) =>
+          // We expect the parent actor to send us a PoisonPill after receiving this message.
+          parent ! MultiPartHtlcFailed(paymentHash, failure, parts.reverse)
+        case d =>
+          log.error(s"unexpected payment failure data ${d.getClass.getSimpleName}")
+      }
+  }
+
+  initialize()
+
+}
+
+object MultiPartPaymentHandler {
+
+  def props(nodeParams: NodeParams, paymentHash: ByteVector32, totalAmount: MilliSatoshi, parent: ActorRef) =
+    Props(new MultiPartPaymentHandler(nodeParams, paymentHash, totalAmount, parent))
+
+  case object PaymentTimeout
+
+  // @formatter:off
+  /** A payment that we're currently holding until we decide to fulfill or fail it. */
+  case class PendingPayment(htlcId: Long, payment: PartialPayment, sender: ActorRef)
+  /** An incoming partial payment. */
+  case class MultiPartHtlc(totalAmount: MilliSatoshi, htlc: UpdateAddHtlc)
+  /** We successfully received all parts of the payment. */
+  case class MultiPartHtlcSucceeded(paymentHash: ByteVector32, parts: List[PendingPayment])
+  /** We aborted the payment because of an inconsistency in the payment set or because we didn't receive the total amount in reasonable time. */
+  case class MultiPartHtlcFailed(paymentHash: ByteVector32, failure: FailureMessage, parts: List[PendingPayment])
+  /** We received an extraneous payment after we reached a final state (succeeded or failed). */
+  case class ExtraHtlcReceived(paymentHash: ByteVector32, payment: PendingPayment, failure: Option[FailureMessage])
+  // @formatter:on
+
+  // @formatter:off
+  sealed trait State
+  case object WAITING_FOR_HTLC extends State
+  case object PAYMENT_SUCCEEDED extends State
+  case object PAYMENT_FAILED extends State
+  // @formatter:on
+
+  // @formatter:off
+  sealed trait Data
```

Comment:
You can factor the `parts` in the trait, and use a `lazy val` to keep the paid amount:

```scala
// @formatter:off
sealed trait Data {
  def parts: List[PendingPayment]
  lazy val paidAmount = parts.map(_.payment.amount).sum
}
case class WaitingForHtlc(parts: List[PendingPayment]) extends Data
case class PaymentSucceeded(parts: List[PendingPayment]) extends Data
case class PaymentFailed(failure: FailureMessage, parts: List[PendingPayment]) extends Data
// @formatter:on
```